Add ephemeral storage support for JavaScript cookies

mrobinson opened the pull request:
This change adds a new class EphemeralStoragePartition which provides
access to an ephemeral version of a RestrictedCookieManager for the
content process. In an upcoming change this will also provide a
StoragePartition for navigation requests in third-party frames in order
to add support for navigation cookies.

<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves brave/brave-browser#12591

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [x] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [x] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

bridiver commented on chromium_src/net/cookies/cookie_monster.h:
```diff
@@ -0,0 +1,19 @@
+#define BRAVE_COOKIE_MONSTER_H                             \
+  std::map<std::string, std::unique_ptr<CookieMonster>>    \
+      ephemeral_cookie_stores_;                            \
+  CookieMonster* GetOrCreateEphemeralCookieStoreForDomain( \
```
this method doesn't need to be in the header, you can pass in ephemeral_cookie_stores_ to non-class method

mrobinson replied:
:+1:

mrobinson replied:
I don't understand why fewer methods is better here, but I'm happy to make this change.

mrobinson replied:
This method also uses net_log_ from `CookieMonster` in order to create the ephemeral monster. Since it's using to instance variables, I think it's simpler for it to be a method. Is there a reason why having fewer methods here is better?

bridiver commented on patches/services-network-public-mojom-cookie_manager.mojom.patch:
```diff
@@ -0,0 +1,24 @@
+ 
+   // Delete session/persistent cookies.
+   CookieDeletionSessionControl session_control = IGNORE_CONTROL;
++
```
never add blank lines or comments to patches

mrobinson replied:
:+1:

bridiver commented on chromium_src/net/base/url_util.cc:
```diff
@@ -80,4 +82,16 @@ bool ParseAuthHostAndPort(base::StringPiece input,
+  // GetDomainAndRegistry might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.empty())
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
```
I don't believe this is the correct behavior. You should create a URL from the domain even if it's empty and in turn Origin will create an opaque origin for those entries

mrobinson replied:
This would mean that all sites that are accessed with IP addresses share the same ephemeral storage. If that the behaviour that we want, I think it makes sense to make that change in a separate PR that also adds tests for that functionality. Since this PR is just about adding support for ephemeral `document.cookie`, I'm not sure it belongs here.

bridiver replied:
no, opaque origins are unique by design, but I think actually it would cause the opposite problem where the same IP in more than one tab would go into different ephemeral storage domains. Since url.Getorigin will return an empty url (and hence create an opaque origin) for file urls, I believe you need to call `SerializeWithNonce`

bridiver replied:
what is `domain` for a file url with this code?

bridiver commented on chromium_src/services/network/restricted_cookie_manager.cc:
```diff
@@ -0,0 +1,32 @@
+  if (url::Origin::Create(url) == top_frame_origin)
+    return;
+  options->ephemeral_storage_domain_ =
+      net::URLToEphemeralStorageDomain(top_frame_origin.GetURL());
```
setting the ephemeral storage domain seems to overly complicate things compared to just setting the url and figuring out the storage domain inside the CookieMonster

mrobinson replied:
:+1:

bridiver commented on chromium_src/net/base/url_util.cc:
```diff
@@ -80,4 +82,16 @@ bool ParseAuthHostAndPort(base::StringPiece input,
   return true;  // Success.
 }
 
+std::string URLToEphemeralStorageDomain(const GURL& url) {
```
I don't think this method is needed here https://github.com/brave/brave-core/pull/7154/files#r524857285

mrobinson replied:
Where do you suggest goes in a way that it can be used in the `CookieMonster` and also in `EphemeralTabStorageHelper` or do you prefer to have two version of this code?

bridiver replied:
I don't think we need two versions of the code if we just pass in the url and only convert it to ephemeral storage domain when we actually create it. The code would only ever be called from one file

mrobinson replied:
This is also called in `EphemeralStorageTabHelper` to manage DOM storage. We need to make sure the conversion from URL to storage domain is in sync.

bridiver commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.h:
```diff
@@ -43,6 +44,7 @@ class EphemeralStorageTabHelper
   friend class content::WebContentsUserData<EphemeralStorageTabHelper>;
   scoped_refptr<content::SessionStorageNamespace> local_storage_namespace_;
   scoped_refptr<content::SessionStorageNamespace> session_storage_namespace_;
+  scoped_refptr<content::TLDEphemeralStorage> tld_ephemeral_storage_;
```
why is this a scoped_refptr?

mrobinson replied:
Reference-counting is used to manage the lifetime of all ephemeral storage. This matches how the lifetime of storage is matched for DOM storage via `SessionStorageNamespace` in upstream Chromium and in our ephemeral storage implementation.

bridiver replied:
ok, I see what's happening now with the weakptr map

bridiver commented on patches/services-network-restricted_cookie_manager.cc.patch:
```diff
@@ -0,0 +1,20 @@
+   //                                 removing deprecation warnings.
+   net_options.set_return_excluded_cookies();
+ 
++  BRAVE_GETALLFORURL
```
this can't work for network cookies so again this doesn't make sense and is not what I described

mrobinson replied:
Based on your comments on Slack @bridiver, it seems you'd like to have `ComputeSameSiteContextForRequest` do the work of storing the top-frame URL into the CookieOptions.
 1. `ComputeSameSiteContextForRequest` isn't called for cookies set via `document.cookie`. Do you imagine it would be called here?
 2. We will be landing support for network cookies in a future PR.

bridiver replied:
as mentioned in slack this was my mistake from searching this page for `BRAVE_SETCANONICALCOOKIE` and not noticing there were two macros with the same name

bridiver commented on patches/net-cookies-cookie_monster.cc.patch:
```diff
@@ -0,0 +1,52 @@
+index 140b61a81dcd748aa22e9dec472f1d375bebcfae..527470acda3863843bca0bcb93e918aa2c35961a 100644
+--- a/net/cookies/cookie_monster.cc
++++ b/net/cookies/cookie_monster.cc
+@@ -579,6 +579,7 @@ void CookieMonster::AttachAccessSemanticsListForCookieList(
```
I think all of these can be eliminated by subclassing CookieMonster and doing a chromium_src override in content::CreateCookieStore

mrobinson replied:
I did consider subclassing `CookieMonster` and maybe that does make sense. The issue is that `content::CreateCookieStore` is not always used to create the `CookieStore`. Crucially, it isn't used for navigation cookies and instead the constructor is called directly. See `net/url_request/url_request_context_builder.cc` and in `services/network/network_context.cc` (using `make_unique`). If we simply override the factory, there is a chance that this will require more work rebasing onto new versions of Chromium.

mrobinson replied:
I chatted on slack with @bridiver about this and he clarified that by using chromium_src override, we should be able to create a subclass called `CookieMonster` and then use macros to rename the original CookieMonster. I will give this a shot.

bridiver commented on content/browser/tld_ephemeral_storage.cc:
```diff
@@ -0,0 +1,90 @@
+      ->id();
+}
+
+TLDEphemeralStorage::TLDEphemeralStorage(TLDEphemeralStorageKey key,
```
why does this need to be in content?

mrobinson replied:
This class is the public interface that `EphemeralTabStorageHelper` uses to access private API in content. Is there a better place for this?

bridiver replied:
what private api?

bridiver replied:
the only private apis I see are accessed outside of this class and we're now adding two new patches for this. It would be better I think to just keep it all included in the browser_context override. Also brave/content should never have been created in the first place and the files that are already in there are misplaced

bridiver replied:
if you created a separate file for tld_ephemeral_storage.cc/.h and included it from browser_context.cc you would put it in chromium_src

mrobinson replied:
There are three interfaces into private API here CreateSessionStorageNamespace and GetSessionStorageNamespaceId, which used to be in `BrowserContext` (still part of content). The final piece is `TLDEphemeralStorage` which talks to the `CookieStore` in order to manage ephemeral cookie lifetime. The idea is that eventually CreateSessionStorageNamespace and GetSessionStorageNamespaceId would become part of `TLDEphemeralStorage`.

mrobinson commented:
@bridiver Okay. I've pushed a new commit, which I believe addresses your concerns. PTAL.

bridiver commented on patches/services-network-public-mojom-cookie_manager.mojom.patch:
```diff
@@ -0,0 +1,20 @@
+   CookieSameSiteContext same_site_cookie_context;
+   bool update_access_time = true;
+   bool return_excluded_cookies = false;
++  url.mojom.Url top_frame_url_for_ephemeral_storage;
```
nit: I don't think we need to specify that this is for ephemeral storage here

bridiver commented on browser/brave_tab_helpers.cc:
```diff
@@ -23,7 +23,7 @@
 #include "brave/components/tor/buildflags/buildflags.h"
 #include "content/public/browser/web_contents.h"
-#include "third_party/blink/public/common/features.h"
+#include "net/base/features.h"
```
looks like you need to add `//net`

mrobinson commented:
Okay. I've pushed a new version of the branch which I believe addresses the review comments above. Thanks!

iefremov commented on chromium_src/content/browser/tld_ephemeral_storage.cc:
```diff
@@ -0,0 +1,72 @@
+    : key_(key), storage_partition_(storage_partition) {
+  DCHECK(active_tld_storage_areas().find(key) ==
+         active_tld_storage_areas().end());
+  active_tld_storage_areas().emplace(key, weak_factory_.GetWeakPtr());
```
I don't quite follow why do we need a weak pointer instead of a raw one?

mrobinson replied:
In this case the weak pointer is just a sanity check which catches errors where the TLDEphemeralStorage is dereferenced and deleted.

iefremov replied:
which checks do you have in mind?

mrobinson replied:
`.get()` on the weak pointer will return nullptr instead of a garbage value, but I will also add a DCHECK to also produce an assertion failure.

iefremov replied:
using `WeakPtr` in this way sounds unusual to me, but probably fine - could you write a comment describing this?

mrobinson replied:
Okay. I'll leave a comment.

iefremov commented on chromium_src/content/browser/tld_ephemeral_storage.cc:
```diff
@@ -0,0 +1,72 @@
+
+#include <map>
+#include "base/no_destructor.h"
+#include "content/browser/dom_storage/dom_storage_context_wrapper.h"
```
there are many unused headers, probably if we drop them we could move the whole class outside of `chromium_src`?

mrobinson replied:
I will definitely remove the unused headers, but the idea is that eventually the private API exposed in browser_context.h can move into this class and it can eventually handle keeping track of all TLD-associated ephemeral DOM storage as well.

iefremov commented on chromium_src/net/cookies/cookie_monster.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef BRAVE_CHROMIUM_SRC_NET_COOKIES_COOKIE_MONSTER_H_
+#define BRAVE_CHROMIUM_SRC_NET_COOKIES_COOKIE_MONSTER_H_
+
+#define CookieMonster SimpleCookieMonster
```
`ChromiumCookieMonster`?

iefremov commented on chromium_src/net/cookies/cookie_monster.cc:
```diff
@@ -0,0 +1,136 @@
+
+#include "net/cookies/cookie_monster.h"
+
+#include <iostream>
```
not needed?

iefremov commented on chromium_src/net/cookies/cookie_monster.cc:
```diff
@@ -0,0 +1,136 @@
+CookieMonster::GetOrCreateEphemeralCookieStoreForTopFrameURL(
+    const GURL& top_frame_url) {
+  std::string domain = URLToEphemeralStorageDomain(top_frame_url);
+  auto it = ephemeral_cookie_stores_.find(domain);
```
no need for this check, `emplace` will do the trick on its own

mrobinson replied:
Won't this result in creating another instantiation of `ChromiumCookieMonster` every time we call this method?

iefremov replied:
yeah, it seems we should leave it alone

iefremov commented on chromium_src/content/public/browser/tld_ephemeral_storage.h:
```diff
@@ -0,0 +1,58 @@
+//
+// TODO(mrobinson): Have this class also manage ephemeral local storage and
+// take care of handing out new instances of session storage.
+class CONTENT_EXPORT TLDEphemeralStorage
```
I'm not sure about future plans for this class, but for now it doesn't look like a "storage", rather something like `TLDEphemeralLifetime`

mrobinson replied:
I haven't changed this yet, because perhaps it doesn't make sense to rename this class and then rename it back in a followup change. I'm happy to keep discussing it though.

iefremov replied:
But even with encapsulated DOMStorage - would it still be "storage"? For now we basically only keep namespaces and manage their lifetime in the tab helper, instead of "storing" something

mrobinson replied:
Okay. I will change the name.

iefremov commented:
Assuming that the patching work is negotiated with @bridiver and most probably cannot be done any better, the whole thing looks good. I'll try to give it test, but for now it failed to build on top of master

iefremov commented:
(I also assume that commit messages will be updated before merging)

mrobinson commented:
> Assuming that the patching work is negotiated with @bridiver and most probably cannot be done any better, the whole thing looks good. I'll try to give it test, but for now it fails to build on top of master

I've pushed a rebased version of this patch and as well a commit addressing review comments.

mrobinson commented:
I've pushed a commit which I believe addresses the remaining review comments.

mrobinson commented:
It seems that the Windows build has an issue with patches to mojom trait headers. I've taken a different tactic, which is to extend `CookieStore` API to add two special methods for ephemeral cookie access. This also has the added benefit of reducing the number of Chromium patches necessary by 2. PTAL.

iefremov reviewed (approved):
pls mention a link to the wiki explainer somewhere in the code (e.g. ephemeral_storage_tab_helper.h and/or ephemeral_storage_lifetime.h)

bridiver commented on chromium_src/content/public/browser/browser_context.h:
```diff
@@ -6,11 +6,14 @@
 
-#define IsOffTheRecord IsTor() const; \
-    virtual bool IsOffTheRecord
+#define IsOffTheRecord \
```
what is this change? Doesn't seem related and also should have 4 space indent for continuation of a line

mrobinson replied:
There have been changes to the lint script which mean that the CI fails unless you do a full format. This will possibly be resolved when https://github.com/brave/brave-core/pull/7402 lands.

mrobinson replied:
In the meantime, I can remove these full format changes.

bridiver commented on chromium_src/content/browser/tld_ephemeral_lifetime.cc:
```diff
@@ -0,0 +1,67 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
this doesn't belong in chromium_src because it's not overriding a chromium file

bridiver commented on chromium_src/content/browser/tld_ephemeral_lifetime.cc:
```diff
@@ -0,0 +1,67 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/chromium_src/content/public/browser/tld_ephemeral_lifetime.h"
```
do not use brave/chromium_src in the include path

bridiver commented on chromium_src/content/browser/browser_context.cc:
```diff
@@ -38,10 +38,11 @@ std::string GetSessionStorageNamespaceId(WebContents* web_contents) {
 }  // namespace content
+
+#include "../../../../content/browser/browser_context.cc"
+#include "brave/chromium_src/content/browser/tld_ephemeral_lifetime.cc"
```
do not use brave/chromium_src in the include path

bridiver commented on chromium_src/content/public/browser/browser_context.h:
```diff
@@ -6,11 +6,14 @@
 #include "../../../../../content/public/browser/browser_context.h"
 #undef IsOffTheRecord
 
+#include "brave/chromium_src/content/public/browser/tld_ephemeral_lifetime.h"
```
don't use brave/chromium_src in the include path

bridiver reviewed (approved):
approved pending the few small fixes

mrobinson commented:
There are three failures in CI:
 * Linux and Android: This is failing in the lint stage (while successful in build and testing). Lint has passed on other platforms and locally. There seem to be issues with the new lint script running on CI and the advice given to me is that this can be ignored if it's unrelated to my change.
 * iOS: It seems that the `AdsWrapperTest testPreferencePersistance` test is failing. This is a known issue and the current advice is to ignore it. Finally, this change should be unrelated to that test failure.

Given this, I feel fairly confident about merging this change despite the CI failures.
